_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+
+The `aaguid` field is [an identifier](https://www.w3.org/TR/webauthn-2/#aaguid) that should be provided by authenticators (but isn't always), which identifies the type of credential used. This can be used to verify the authenticator's make and model. Also, it can be useful for denying access for outdated authenticators with known security vulnerabilities.
+
+The credential data class has one constructor that uses the Yubico [`RegistrationResult`](https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core-minimal/latest/com/yubico/webauthn/RegistrationResult.html) and the [`AuthenticatorAttestationResponse`](https://www.w3.org/TR/webauthn/#iface-authenticatorattestationresponse) object (converted from JavaScript to Java), along with a username and name for the credential.
+
+```java
+public Authenticator(RegistrationResult result, 
+                     AuthenticatorAttestationResponse response, 
+                     AppUser user, 
+                     String name) {
+    Optional<AttestedCredentialData> attestationData = response.getAttestation()
+        .getAuthenticatorData()
+        .getAttestedCredentialData();
+    this.credentialId = result.getKeyId().getId();
+    this.publicKey = result.getPublicKeyCose();
+    this.aaguid = attestationData.get().getAaguid();
+    this.count = result.getSignatureCount();
+    this.name = name;
+    this.user = user;
+}
+```
+
+Create another `CrudRepository` to mange the `Authenticator` objects.
+
+```java
+import java.util.List;
+import java.util.Optional;
+
+import com.yubico.webauthn.data.ByteArray;
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface AuthenticatorRepository extends CrudRepository<Authenticator, Long> {
+    Optional<Authenticator> findByCredentialId(ByteArray credentialId);
+    List<Authenticator> findAllByUser (AppUser user);
+    List<Authenticator> findAllByCredentialId(ByteArray credentialId);
+}
+```
+
+## Application properties
+
+Create a configuration bean to hold the application's properties:
+
+```java
+import java.util.Set;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+import lombok.Getter;
+import lombok.Setter;
+
+@Configuration
+@Getter
+@Setter
+@ConfigurationProperties(prefix = "authn")
+public class WebAuthProperties {
+    private String hostName;
+    private String display;
+    private Set<String> origin;
+}
+```
+
+Add configuration for localhost in your `src/main/resources/application.properties`:
+
+```properties
+authn.hostname=localhost
+authn.display=Otka WebAuthn Sample Application
```

Comment:
```suggestion
authn.display=Okta WebAuthn Sample Application
```

Does this app use Okta? If not, we might want to name it `Spring Boot WebAuthn Sample Application`